### PR TITLE
feat: add sign blob commands

### PR DESCRIPTION
**Pull request by rrsemlani**
Adding sign blob commands.
Spec Pr used for ref: https://github.com/notaryproject/notation/pull/811

### CODE REVIEW

File: cmd/notation/main.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ func main() {
 	}
 	cmd.AddCommand(
 		signCommand(nil),
+		blobSignCommand(nil),
```
Comment (rgnote) [inline cmd/notation/main.go:38]: nit: put all blob commends at one place for readability

File: cmd/notation/blob.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+import (
+	"errors"
+	"fmt"
+	"github.com/notaryproject/notation/cmd/notation/internal/experimental"
+	"github.com/notaryproject/notation/internal/cmd"
+	"github.com/spf13/cobra"
+)
+
+func blobSignCommand(opts *signOpts) *cobra.Command {
```
Comment (rgnote) [inline cmd/notation/blob.go:24]: `signOpts` defined in `sign.go` has more flags than what is needed for `notation blob sign` command. Can we have a separate `signOpts` for blob signing?

File: cmd/notation/blob.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	if opts == nil {
+		opts = &signOpts{
+			inputType: inputTypeRegistry, // remote registry by default
```
Comment (rgnote) [inline cmd/notation/blob.go:27]: Registries do not apply to blobs. Lets remove this.

File: cmd/notation/blob.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+		}
+	}
+	longMessage := `Sign artifacts
```
Comment (rgnote) [inline cmd/notation/blob.go:30]: Can you use the log message as is from `notation blob sign` section in https://github.com/notaryproject/notation/pull/811/files?diff=unified&w=0?

File: cmd/notation/blob.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+Example - Sign a BLOB artifact identified by a tag (Notation will resolve tag to digest)
+  notation blob sign <registry>/<repository>:<tag>
+
+Example - Sign a BLOB artifact stored in a registry and specify the signature expiry duration, for example 24 hours
+  notation blob sign --expiry 24h <registry>/<repository>@<digest>
+`
+
+	command := &cobra.Command{
+		Use:   "blob sign [flags] <reference>",
```
Comment (rgnote) [inline cmd/notation/blob.go:54]: reference?

File: cmd/notation/blob.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+func blobInspectCommand(opts *inspectOpts) *cobra.Command {
+	if opts == nil {
+		opts = &inspectOpts{}
+	}
+	longMessage := `Inspect all signatures associated with the signed artifact.
+
+Example - Inspect signatures on an BLOB artifact identified by a digest:
+  notation blob inspect <registry>/<repository>@<digest>
```
Comment (rgnote) [inline cmd/notation/blob.go:85]: ?

File: cmd/notation/blob.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
+// Copyright The Notary Project Authors.
```
Comment (priteshbandi) [inline cmd/notation/blob.go:1]: Please move this inside `blob` directory similar to cert, plugin etc. Also, rename file to sign.go.

File: cmd/notation/blob.go [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+	"github.com/spf13/cobra"
+	"os"
+	"strings"
+	"time"
+)
+
+type blobOpts struct {
+	cmd.LoggingFlagOpts
+	cmd.SignerFlagOpts
+	SecureFlagOpts
```
Comment (priteshbandi) [inline cmd/notation/blob.go:37]: Do we need `SecureFlagOpts` ?

File: cmd/notation/blob.go [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+	expiry        time.Duration
+	desc          ocispec.Descriptor
+	sigRepo       registry.Repository
+	pluginConfig  []string
+	userMetadata  []string
+	blobPath      string
+	signaturePath string
```
Comment (priteshbandi) [inline cmd/notation/blob.go:44]:
```suggestion
	signatureDirectory string
```

File: cmd/notation/blob.go [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+  notation blob sign --expiry 24h <blob_path>
+`
+
+	command := &cobra.Command{
+		Use:   "blob sign [flags] <blobPath>",
+		Short: "Sign BLOB artifacts",
+		Long:  longMessage,
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return errors.New("missing blob_path")
```
Comment (priteshbandi) [inline cmd/notation/blob.go:90]: What is `blob_path` ? How will user interpret it?

File: cmd/notation/blob.go [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+	"time"
+)
+
+type blobOpts struct {
+	cmd.LoggingFlagOpts
+	cmd.SignerFlagOpts
+	SecureFlagOpts
+	expiry        time.Duration
+	desc          ocispec.Descriptor
+	sigRepo       registry.Repository
```
Comment (priteshbandi) [inline cmd/notation/blob.go:40]: Why do we need this ?

File: cmd/notation/blob.go [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+	pluginConfig  []string
+	userMetadata  []string
+	blobPath      string
+	signaturePath string
+	outputFormat  string
```
Comment (priteshbandi) [inline cmd/notation/blob.go:45]: are we missing option for some flags like `--force` ?

File: cmd/notation/blob.go [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+func blobSignCommand(opts *blobOpts) *cobra.Command {
+	if opts == nil {
+		opts = &blobOpts{}
```
Comment (priteshbandi) [inline cmd/notation/blob.go:50]: Is this correct ?

File: cmd/notation/blob.go [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+			SignatureMediaType: mediaType,
+			ExpiryDuration:     opts.expiry,
+			PluginConfig:       pluginConfig,
+		},
+		UserMetadata: userMetadata,
+	}
+	return blobOpts, nil
+}
+
+func blobInspectCommand(opts *blobOpts) *cobra.Command {
```
Comment (priteshbandi) [inline cmd/notation/blob.go:163]: Please create new .go file for inspect.

File: cmd/notation/blob.go [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+	// set log level
+	ctx := opts.LoggingFlagOpts.InitializeLogger(command.Context())
+
+	if opts.outputFormat != cmd.OutputJSON && opts.outputFormat != cmd.OutputPlaintext {
+		return fmt.Errorf("unrecognized output format %s", opts.outputFormat)
+	}
+
+	output := inspectOutput{MediaType: opts.desc.MediaType, Signatures: []signatureOutput{}}
+
+	sigBlob, _, _ := opts.sigRepo.FetchSignatureBlob(ctx, opts.desc)
```
Comment (priteshbandi) [inline cmd/notation/blob.go:208]: we dont have sigRepo in case of blob signing.

File: cmd/notation/notation
Comment (Two-Hearts) [inline cmd/notation/notation:1]: nit:
We should not include the go build `notation` binary in the PR.

File: cmd/notation/blob/cmd.go [hunk 16]
```diff
@@ -0,0 +1 @@
+package blob
```
Comment (priteshbandi) [inline cmd/notation/blob/cmd.go:1]: Please remove empty files.

File: cmd/notation/notation
Comment (priteshbandi) [inline cmd/notation/notation:1]: Please remove binary

File: cmd/notation/blob/sign.go [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+
+func runBlobSign(command *cobra.Command, cmdOpts *blobSignOpts) error {
+	// set log level
+	ctx := cmdOpts.LoggingFlagOpts.InitializeLogger(command.Context())
+
+	// initialize
+	signer, err := cmd.GetSigner(ctx, &cmdOpts.SignerFlagOpts)
+	if err != nil {
+		return err
+	}
```
Comment (priteshbandi) [inline cmd/notation/blob/sign.go:105]: Add a TODO that we will need to replace signer with actual blob signer implementation in notation-go

File: cmd/notation/blob/cmd.go (hunk 16 above)
Comment (rrsemlani) [reply cmd/notation/blob/cmd.go:1]: Will remove them

File: cmd/notation/main.go [hunk 18]
```diff
@@ -45,6 +46,7 @@ func main() {
 		logoutCommand(nil),
 		versionCommand(),
 		inspectCommand(nil),
+		blob.SignCommand(nil),
```
[the remark this answers is not in the document]
Comment (rrsemlani) [reply cmd/notation/main.go:49]: Will update in new rev

File: cmd/notation/notation
Comment (rrsemlani) [reply cmd/notation/notation:1]: Will remove in new rev

File: cmd/notation/blob/sign.go [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+	return command
+}
+
+func runBlobSign(command *cobra.Command, cmdOpts *blobSignOpts) error {
+	// set log level
+	ctx := cmdOpts.LoggingFlagOpts.InitializeLogger(command.Context())
+
+	// Todo: we will need to replace signer with actual blob signer implementation in notation-go
+	// initialize
+	signer, err := cmd.GetSigner(ctx, &cmdOpts.SignerFlagOpts)
```
Comment (priteshbandi) [inline cmd/notation/blob/sign.go:105]: `cmd.GetSigner` returns `notation.Signer` which doesnt supports for BlobSigning.

File: cmd/notation/blob/sign.go [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+	if err != nil {
+		return notation.SignOptions{}, err
+	}
+	blobOpts := notation.SignOptions{
+		SignerSignOptions: notation.SignerSignOptions{
+			SignatureMediaType: mediaType,
+			ExpiryDuration:     opts.expiry,
+			PluginConfig:       pluginConfig,
+		},
+		UserMetadata: userMetadata,
```
Comment (priteshbandi) [inline cmd/notation/blob/sign.go:145]: Please use blob related opts

File: cmd/notation/blob/sign.go [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+	// initialize
+	signer, err := cmd.GetSigner(ctx, &cmdOpts.SignerFlagOpts)
+	if err != nil {
+		return err
+	}
+	blobOpts, err := prepareBlobSigningOpts(ctx, cmdOpts)
+	if err != nil {
+		return err
+	}
+	contents, err := os.ReadFile(cmdOpts.blobPath)
```
Comment (priteshbandi) [inline cmd/notation/blob/sign.go:113]: Dont read file; pass the reader to SignBlob.

File: cmd/notation/blob/sign.go [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+	if err != nil {
+		return err
+	}
+	// core process
+	_, _, err = notation.SignBlob(ctx, signer, strings.NewReader(string(contents)), blobOpts) //PlaceHolder
```
Comment (priteshbandi) [inline cmd/notation/blob/sign.go:118]: Can we please add implementation to write signature file?

**Pull request closed without merging**